pertpy/tools/_differential_gene_expression/_base.py

```diff
@@ -449,20 +506,437 @@ def _map_genes_categories_highlight(
+            ... )
+            >>> edgr.plot_paired(pdata, var_names=res_df["variable"][:8], groupby="Treatment", pairedby="Major celltype")
+
+        Preview:
+            .. image:: /_static/docstring_previews/de_paired_expression.png
+        """
+        if boxplot_properties is None:
+            boxplot_properties = {}
+        groups = adata.obs[groupby].unique()
+        if len(groups) != 2:
+            raise ValueError("The number of groups in the group_by column must be exactly 2 to enable paired testing")
+        if pairedby is not None and hue is not None and (adata.obs.groupby(pairedby)[hue].nunique().max() > 1):
+            raise ValueError("Each paired sample must have an unambiguous hue")
+
+        adata = adata[:, var_names]
+
+        if pairedby is not None and any(adata.obs[[groupby, pairedby]].value_counts() > 1):
+            logger.info("Performing pseudobulk for paired samples")
+            ps = PseudobulkSpace()
+            adata = ps.compute(
+                adata, target_col=groupby, groups_col=pairedby, layer_key=layer, mode="sum", min_cells=1, min_counts=1
+            )
+
+        if layer is not None:
+            X = adata.layers[layer]
+        else:
+            X = adata.X
+        try:
+            X = X.toarray()
+        except AttributeError:
+            pass
+
+        groupby_cols = [groupby]
+        if pairedby is not None:
+            groupby_cols.insert(0, pairedby)
+        if hue is not None:
+            groupby_cols.insert(0, hue)
+
+        df = adata.obs.loc[:, groupby_cols].join(pd.DataFrame(X, index=adata.obs_names, columns=var_names))
+
+        if pairedby is not None:
+            # remove unpaired samples
+            paired_samples = set(df[df[groupby] == groups[0]][pairedby]) & set(df[df[groupby] == groups[1]][pairedby])
+            df = df[df[pairedby].isin(paired_samples)]
+            removed_samples = adata.obs[pairedby].nunique() - len(df[pairedby].unique())
+            if removed_samples > 0:
+                logger.warning(f"{removed_samples} unpaired samples removed")
+
+            # perform statistics (paired ttest)
+            if pvalues is None:
+                _, pvalues = scipy.stats.ttest_rel(
+                    df.loc[
+                        df[groupby] == groups[0],
+                        var_names,
+                    ],
+                    df.loc[
+                        df[groupby] == groups[1],
+                        var_names,
+                    ],
+                )
+
+            df.reset_index(drop=False, inplace=True)
+
+        else:
+            if pvalues is None:
+                _, pvalues = scipy.stats.ttest_ind(
+                    df.loc[
+                        df[groupby] == groups[0],
+                        var_names,
+                    ],
+                    df.loc[
+                        df[groupby] == groups[1],
+                        var_names,
+                    ],
+                )
```

Comment:
Since the tools for proper DE (also including t-test) are now included in pertpy, I'd favor referring to a data frame with precomputed DE results here, rather than doing the statistical test in the plotting function.

Reply:
@Zethson and I also discussed at some point that it could make sense to have a `DEResults` that contains the plotting functions rather than having the plotting functions in the model class. In that state of the results object would already contain the p-values.

Reply:
> @Zethson and I also discussed at some point that it could make sense to have a DEResults that contains the plotting functions rather than having the plotting functions in the model class. In that state of the results object would already contain the p-values.

That's something I would consider adapting in a follow up PR. We just wanted to get this ready for the conference.

Reply:
> Since the tools for proper DE (also including t-test) are now included in pertpy, I'd favor referring to a data frame with precomputed DE results here, rather than doing the statistical test in the plotting function.

Totally agree, however, the "typical" DE results won't necessarily contain the results of paired samples, which, however, is required for this plot. So, I'm not sure how to implement that with only the `results_df` (same problem as for the `plot_fold_change` method).

Reply:
> If I understand you correctly, that's what I'm currently doing. I'm just not sure how to incorporate the pairing into this based on the results_df alone, without calculating a lot of new metrics. But maybe the current plot version is sufficient for this PR, and we can consider adding the one including paired samples as a separate plot in a future PR.

At least not for the `plot_paired` though? It doesn't take the results df as input. 

> Totally agree, however, the "typical" DE results won't necessarily contain the results of paired samples, which, however, is required for this plot. 

Well, that's up to the user to use the appropriate model definition. I think I'd go with something like

```python
edgr = pt.tl.EdgeR(pdata, design="~Treatment+Patient")
edgr.fit()
res_df = edgr.test_contrasts(
     edgr.contrast(column="Treatment", baseline="Chemo", group_to_compare="Anti-PD-L1+Chemo")
)
edgr.plot_paired(
   pdata, 
   # specifying the whole dataframe and selected `var_names` separately could also be an option. 
   de_res = res_df[:10], 
   groupby="Treatment",
   pairedby="Patient"
)
```

Since the "pairedby" variable is part of the model definition, the p-values are appropriate for paired measurements. 

I'd then just replace the part that calculates the p-values with obtaining them from the dataframe.

Reply:
Ok, got it - thank you so much for your input! I've implemented it that way now.